File: erlang/mapReduce.erl
-module(mapreduce).
-export([start/0, map_reduce/1]).

% Step 1: Map - Tokenize each document and emit {word, 1} pairs
map_function(Document) ->
    Words = string:tokens(string:to_lower(Document), " "),
    [{Word, 1} || Word <- Words].

% Step 2: Shuffle and Sort - Sort by the key (word)
sort_results(Results) ->
    lists:keysort(1, Results).

% Step 3: Reduce - Aggregate counts for each word
reduce_function({Word, Counts}) ->
    {Word, lists:sum(Counts)}.

% Worker process for mapping
map_worker(Documents) ->
    MappedResults = lists:flatten(lists:map(fun map_function/1, Documents)),
    SortedResults = sort_results(MappedResults),
    {self(), SortedResults}.

% Worker process for reducing
reduce_worker(SortedResults) ->
    ReducedResults = lists:map(fun reduce_function/1, SortedResults),
    {self(), ReducedResults}.

% Master process coordinating the MapReduce
master(Documents) ->
    % Split documents among map workers
    MapWorkers = [spawn(mapreduce, map_worker, [Chunk]) || Chunk <- lists:split(length(Documents) div 2, Documents)],

    % Collect mapped results from map workers
    MappedResults = lists:flatten([receive {Worker, Results} -> Results end || Worker <- MapWorkers]),

    % Shuffle and Sort
    SortedResults = sort_results(MappedResults),

    % Split sorted results among reduce workers
    ReduceWorkers = [spawn(mapreduce, reduce_worker, [Chunk]) || Chunk <- lists:split(length(SortedResults) div 2, SortedResults)],

    % Collect reduced results from reduce workers
    ReducedResults = lists:flatten([receive {Worker, Results} -> Results end || Worker <- ReduceWorkers]),

    % Display the final word count
    io:format("Word Count: ~p~n", [ReducedResults]).

% Entry point
start() ->
    Documents = ["Hello world",
                 "Erlang is awesome",
                 "Hello Erlang",
                 "MapReduce in Erlang",
                 "Erlang programming"],
    master(Documents).

File: elixir/mapReduce.ex
defmodule MapReduceExample do
  # Step 1: Map - Tokenize each document and emit {word, 1} pairs
  def map_function(document) do
    document
    |> String.downcase()
    |> String.split(~r/\s+/)
    |> Enum.map(fn word -> {word, 1} end)
  end

  # Step 2: Shuffle and Sort - Sort by the key (word)
  def sort_results(results) do
    Enum.sort(results, fn {word1, _}, {word2, _} -> word1 <= word2 end)
  end

  # Step 3: Reduce - Aggregate counts for each word
  def reduce_function({word, counts}) do
    {word, Enum.sum(counts)}
  end

  # Worker process for mapping
  def map_worker(documents) do
    mapped_results =
      documents
      |> Enum.flat_map(&map_function/1)
    sort_results(mapped_results)
  end

  # Worker process for reducing
  def reduce_worker(sorted_results) do
    grouped_results =
      sorted_results
      |> Enum.group_by(fn {word, _} -> word end)
    Enum.map(grouped_results, &reduce_function/1)
  end

  # Master process coordinating the MapReduce
  def master(documents) do
    # Split documents among map workers
    {map_chunk1, map_chunk2} = Enum.split(documents, div(Enum.count(documents), 2))
    map_workers = [map_chunk1, map_chunk2] |> Enum.map(&map_worker/1) |> Enum.flatten()

    # Shuffle and Sort
    sorted_results = sort_results(map_workers)

    # Split sorted results among reduce workers
    {reduce_chunk1, reduce_chunk2} = Enum.split(sorted_results, div(Enum.count(sorted_results), 2))
    reduce_workers = [reduce_chunk1, reduce_chunk2] |> Enum.map(&reduce_worker/1) |> Enum.flatten()

    # Display the final word count
    reduce_workers
  end
end

# Example usage
documents = [
  "Hello world",
  "Elixir is awesome",
  "Hello Elixir",
  "MapReduce in Elixir",
  "Elixir programming"
]

final_result = MapReduceExample.master(documents)
IO.puts("Word Count:")
Enum.each(final_result, &IO.inspect/1)
